### PR TITLE
fix warning and background block problem

### PR DESCRIPTION
**Pull request by sherry-1001**
Signed-off-by: zhichenj <zhichen.jiang@intel.com>

### What problem does this PR solve?
1. Fix warning about no return type.
2. Workaround to solve block problem for background cleaner.

Tests 

- Unit test
- Integration test
- No code

Side effects
   No measure

### CODE REVIEW

**Review by ZiyanShi — APPROVED** [review]

**Comment by ZiyanShi** [comment]
> Signed-off-by: zhichenj [zhichen.jiang@intel.com](mailto:zhichen.jiang@intel.com)
> 
> ### What problem does this PR solve?
> 1. Fix warning about no return type.
> 2. Workaround to solve block problem for background cleaner.
> 
> Tests
> 
> * Unit test
> * Integration test
> * No code
> 
> Side effects
> 
> * Performance regression
>   
>   * Consumes more CPU
>   * Consumes more MEM
>   * Consumes more PMEM



> Signed-off-by: zhichenj [zhichen.jiang@intel.com](mailto:zhichen.jiang@intel.com)
> 
> ### What problem does this PR solve?
> 1. Fix warning about no return type.
> 2. Workaround to solve block problem for background cleaner.
> 
> Tests
> 
> * Unit test
> * Integration test
> * No code
> 
> Side effects
> 
> * Performance regression
>   
>   * Consumes more CPU
>   * Consumes more MEM
>   * Consumes more PMEM

That's a horrible fix.
A bug fix caused KVDK to consumer more CPU, VMem and PMem?
Take some time to review your PR before actually submit.

**Comment by ZiyanShi** [comment]
> Signed-off-by: zhichenj [zhichen.jiang@intel.com](mailto:zhichen.jiang@intel.com)
> 
> ### What problem does this PR solve?
> 1. Fix warning about no return type.
> 2. Workaround to solve block problem for background cleaner.
> 
> Tests
> 
> * Unit test
> * Integration test
> * No code
> 
> Side effects
> 
> * Performance regression
>   
>   * Consumes more CPU
>   * Consumes more MEM
>   * Consumes more PMEM



> Signed-off-by: zhichenj [zhichen.jiang@intel.com](mailto:zhichen.jiang@intel.com)
> 
> ### What problem does this PR solve?
> 1. Fix warning about no return type.
> 2. Workaround to solve block problem for background cleaner.
> 
> Tests
> 
> * Unit test
> * Integration test
> * No code
> 
> Side effects
> 
> * Performance regression
>   
>   * Consumes more CPU
>   * Consumes more MEM
>   * Consumes more PMEM

That's a horrible fix.
A bug fix caused KVDK to consumer more CPU, VMem and PMem?
Take some time to review your PR before actually submit.

File: engine/kv_engine.cpp [hunk 1]
```diff
@@ -2052,27 +2052,16 @@ void KVEngine::delayFree(const OldDeleteRecord& old_delete_record) {
...
-    {
-      std::unique_lock<SpinMutex> ul(bg_work_signals_.terminating_lock);
-      if (!bg_work_signals_.terminating) {
-        bg_work_signals_.old_records_cleaner_cv.wait_for(ul, interval);
-      }
-    }
-    bg_cleaner_processing_ = true;
-    CleanExpired();
-    old_records_cleaner_.TryGlobalClean();
+  while (!closing_) {
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:2063]: !bg_work_signals_.termination?

File: tests/tests.cpp [hunk 2]
```diff
@@ -2026,7 +2026,9 @@ TEST_F(EngineBasicTest, TestBackGroundCleaner) {
       int64_t ttl_time;
       Status s = engine->GetTTL(key, &ttl_time);
       if (s == Status::Ok) {
-        ASSERT_EQ(INT32_MAX / 10000, ttl_time / 10000);
+        if (ttl_time <= 1) {
+          ASSERT_EQ(INT32_MAX / 10000, ttl_time / 10000);
```
Comment (ZiyanShi) [inline tests/tests.cpp:2030]: That's a strange ASSERT_EQ.
`INT32_MAX/10000 > 0 >= ttl_time/10000`

Comment (ZiyanShi) [inline tests/tests.cpp:2030]: That's a strange ASSERT_EQ.
`INT32_MAX/10000 > 0 >= ttl_time/10000`

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

File: engine/kv_engine.cpp [hunk 3]
```diff
@@ -2037,9 +2036,9 @@ void KVEngine::delayFree(const OldDataRecord& old_data_record) {
   old_records_cleaner_.PushToCache(old_data_record);
   // To avoid too many cached old records pending clean, we try to clean cached
   // records while pushing new one
-  if (old_records_cleaner_.NumCachedOldRecords() > kMaxCachedOldRecords &&
-      !bg_cleaner_processing_) {
-    bg_work_signals_.old_records_cleaner_cv.notify_all();
+  if (!need_clean &&
```
Comment (JiayuZzz) [inline engine/kv_engine.cpp:2039]: need_clean_ as private. Or need_global_clean_?

File: engine/version/old_records_cleaner.cpp [hunk 4]
```diff
@@ -36,7 +36,7 @@ void OldRecordsCleaner::PushToCache(const OldDeleteRecord& old_delete_record) {
   tc.old_delete_records.emplace_back(old_delete_record);
 }
 
-void OldRecordsCleaner::PushToGloble(
+void OldRecordsCleaner::PushToGlabal(
```
Comment (JiayuZzz) [inline engine/version/old_records_cleaner.cpp:39]: Typo. Glabal -> Global

**Review by JiayuZzz — APPROVED** [review]

File: engine/kv_engine.cpp [hunk 5]
```diff
@@ -2154,13 +2144,24 @@ void KVEngine::CleanExpired() {
       }
       bucket_iter++;
     }
-    old_records_cleaner_.PushToGloble(expired_record_queue);
+
+    if (!expired_record_queue.empty() &&
+        expired_record_queue.size() % kMaxCachedOldRecords) {
+      old_records_cleaner_.PushToGlobal(expired_record_queue);
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:2150]: May be never reached.

File: tests/tests.cpp [hunk 6]
```diff
@@ -2491,6 +2424,81 @@ TEST_F(EngineBasicTest, TestHashTableRangeIter) {
...
+    }
+  };
+  auto ExpiredClean = [&]() {
+    auto test_kvengine = static_cast<KVEngine*>(engine);
+    test_kvengine->CleanOutDated();
+  };
+
+  auto ExpireString = [&](Status s) {
+    for (int i = 0; i < cnt; ++i) {
+      std::string key = std::to_string(i) + "stringk";
```
Comment (ZiyanShi) [inline tests/tests.cpp:2456]: This Status s parameter, is only Status::Ok.

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

File: engine/kv_engine.cpp [hunk 7]
```diff
@@ -2154,13 +2144,24 @@ void KVEngine::CleanExpired() {
...
+    }
+    if (std::chrono::duration_cast<std::chrono::seconds>(
+            std::chrono::system_clock::now() - start_ts)
+                .count() > interval ||
+        need_clean_records_) {
+      need_clean_records_ = true;
+      old_records_cleaner_.TryGlobalClean();
+      need_clean_records_ = false;
+      start_ts = std::chrono::system_clock::now();
+    }
```
Comment (JiayuZzz) [inline engine/kv_engine.cpp:2161]: Put remaining expired records to global

File: engine/kv_engine.cpp [hunk 8]
```diff
@@ -2154,13 +2144,24 @@ void KVEngine::CleanExpired() {
       }
       bucket_iter++;
     }
-    old_records_cleaner_.PushToGloble(expired_record_queue);
+
+    if (!expired_record_queue.empty() &&
+        expired_record_queue.size() % kMaxCachedOldRecords) {
+      old_records_cleaner_.PushToGlobal(expired_record_queue);
+      expired_record_queue.clear();
+    }
```
Comment (JiayuZzz) [inline engine/kv_engine.cpp:2152]: Put remaining expired records to global

**Review by JiayuZzz — APPROVED** [review]
LGTM

**Review by ZiyanShi — APPROVED** [review]

**Pull request merged**